### PR TITLE
specs for handlePayment

### PR DESCRIPTION
**Pull request by yzhang90**

### CODE REVIEW

File: gnosis/spec-tmpl.k [hunk 1]
```diff
@@ -106,16 +106,11 @@
...
      andBool #rangeAddress(#ORIGIN_ID)
-     andBool #rangeUInt(256, MASTER_BAL)
-     andBool #rangeUInt(256, PROXY_BAL)
-     andBool #rangeUInt(256, ORIGIN_BAL)
      andBool #rangeUInt(256, NOW)
      andBool #rangeUInt(128, BLOCK_NUM) // Solidity
-     //
-     andBool #MASTER_COPY_ID =/=Int #PROXY_ID
-     andBool #MASTER_COPY_ID =/=Int #ORIGIN_ID
-     andBool #PROXY_ID       =/=Int #ORIGIN_ID
```
Comment (daejunpark) [inline gnosis/spec-tmpl.k]: why did you remove these?

File: gnosis-imap/gnosis-imap-spec.ini [hunk 2]
```diff
@@ -148,6 +153,106 @@ proxy_storage:
...
+pc: 19729 => 20290
+
+[handlePayment-send-success-origin]
+gas: #gas(INITGAS, NONMEMGAS, MEMGAS) => #gas(INITGAS, NONMEMGAS +Int #sendGas(BYZANTIUM, #ORIGIN_ID, TOTAL_AMOUNT, true) +Int 618, MEMGAS +Int (Cmem(BYZANTIUM, FINAL_MEM_USAGE) -Int Cmem(BYZANTIUM, MU)))
+proxy_bal: PROXY_BAL => PROXY_BAL -Int TOTAL_AMOUNT
+origin_bal: ORIGIN_BAL => ORIGIN_BAL +Int TOTAL_AMOUNT
++requires:
+    andBool #REFUND_RECEIVER ==Int 0
+    andBool #sendSuccess(#ORIGIN_ID)
+    andBool #PROXY_ID =/=K #ORIGIN_ID
```
Comment (daejunpark) [inline gnosis-imap/gnosis-imap-spec.ini]: TODO: #PROXY_ID ==K #ORIGIN_ID

File: gnosis-imap/gnosis-imap-spec.ini [hunk 3]
```diff
@@ -148,6 +153,106 @@ proxy_storage:
...
+            <storage> _ </storage>
+            <origStorage> _ </origStorage>
+            <nonce> _ </nonce>
+          </account>
++requires:
+    andBool 0 <Int #REFUND_RECEIVER andBool (notBool #REFUND_RECEIVER in #precompiledAccounts(BYZANTIUM))
+    andBool #rangeUInt(256, RECEIVER_BAL)
+
+    andBool #sendSuccess(#REFUND_RECEIVER)
+    andBool #PROXY_ID =/=K #REFUND_RECEIVER
```
Comment (daejunpark) [inline gnosis-imap/gnosis-imap-spec.ini]: TODO: #PROXY_ID ==K #REFUND_RECEIVER

File: gnosis/abstract-semantics.k [hunk 4]
```diff
@@ -41,9 +46,47 @@ module ABSTRACT-SEMANTICS
...
+         <balance> ORIGFROM => ORIGFROM -Int VALUE </balance>
+         ...
+       </account>
+       <account>
+         <acctID> ACCTTO </acctID>
+         <balance> ORIGTO => ORIGTO +Int VALUE </balance>
+         ...
+       </account>
+    requires #sendSuccess(ACCTTO)
+     andBool ACCTFROM =/=K ACCTTO
```
Comment (daejunpark) [inline gnosis/abstract-semantics.k]: TODO: ACCTFROM ==K ACCTTO

File: gnosis-imap/gnosis-imap-spec.ini [hunk 5]
```diff
@@ -148,14 +153,114 @@ proxy_storage:
...
+wordStack: #REFUND_RECEIVER : GAS_TOKEN : GAS_PRICE : DATA_GAS : START_GAS : RETURN_LOC : WS => RETURN_LOC : WS
+pc: 19729 => 20290
+
+[handlePayment-send-success-origin]
+gas: #gas(INITGAS, NONMEMGAS, MEMGAS) => #gas(INITGAS, NONMEMGAS +Int #sendGas(BYZANTIUM, #ORIGIN_ID, TOTAL_AMOUNT, true) +Int 618, MEMGAS +Int (Cmem(BYZANTIUM, FINAL_MEM_USAGE) -Int Cmem(BYZANTIUM, MU)))
+proxy_bal: PROXY_BAL => PROXY_BAL -Int TOTAL_AMOUNT
+origin_bal: ORIGIN_BAL => ORIGIN_BAL +Int TOTAL_AMOUNT
++requires:
+    andBool #REFUND_RECEIVER ==Int 0
+    andBool #sendSuccess(#ORIGIN_ID)
```
Comment (daejunpark) [inline gnosis-imap/gnosis-imap-spec.ini]: could you explain why this `#sendSuccess(#ORIGIN_ID)` works at this time?

File: gnosis-imap/gnosis-imap-spec.ini [hunk 6]
```diff
@@ -148,14 +153,114 @@ proxy_storage:
...
+          </account>
++requires:
+    andBool 0 <Int #REFUND_RECEIVER andBool (notBool #REFUND_RECEIVER in #precompiledAccounts(BYZANTIUM))
+    andBool #rangeUInt(256, RECEIVER_BAL)
+
+    andBool #sendSuccess(#REFUND_RECEIVER)
+    andBool #PROXY_ID =/=K #REFUND_RECEIVER
+    andBool TOTAL_AMOUNT <=Int PROXY_BAL
+    andBool RECEIVER_BAL +Int TOTAL_AMOUNT <Int pow256
+
```
Comment (daejunpark) [inline gnosis-imap/gnosis-imap-spec.ini]: don't we need `[handlePayment-send-failure]`?

**Comment by daejunpark** [comment]
@yzhang90 overall looks good. The three TODO comments are for the reference. Please add it as a comment so that we do not forget later (in the next PR).

File: gnosis/spec-tmpl.k (hunk 1 above)
Comment (yzhang90) [reply gnosis/spec-tmpl.k]: These constraints are mainly used for transferFunds. I think we'd better move these to each spec that involves transferFunds instead of putting it in the template.

File: gnosis-imap/gnosis-imap-spec.ini (hunk 2 above)
Comment (daejunpark) [reply gnosis-imap/gnosis-imap-spec.ini]: seems not possible
